### PR TITLE
Add timeout parameter to IPCProvider class

### LINKED ISSUE #309: IPCProvider timeout paramater
* Version: 3.13.5
* Python: 2.7
* OS: Ubuntu 16.04

### What was wrong?
A request has timed out.

Please include any of the following that are applicable:
* The code which produced the error
`web3.manager.request_blocking('parity_pendingTransactions', [])`
* The full output of the error
`
Traceback (most recent call last):
  File "/home/user/clone/mempool/eth.py", line 78, in get_mempool
    'parity_pendingTransactions', []
  File "/home/user/.virtualenvs/prune/local/lib/python2.7/site-packages/web3/manager.py", line 112, in request_blocking
    response = self._make_request(method, params)
  File "/home/user/.virtualenvs/prune/local/lib/python2.7/site-packages/web3/manager.py", line 95, in _make_request
    return make_request_fn(method, params)
  File "/home/user/.virtualenvs/prune/local/lib/python2.7/site-packages/web3/middleware/attrdict.py", line 20, in middleware
    response = make_request(method, params)
  File "/home/user/.virtualenvs/prune/local/lib/python2.7/site-packages/web3/middleware/formatting.py", line 25, in middleware
    response = make_request(method, params)
  File "/home/user/.virtualenvs/prune/local/lib/python2.7/site-packages/web3/providers/ipc.py", line 103, in make_request
    timeout.sleep(0)
  File "/home/user/.virtualenvs/prune/local/lib/python2.7/site-packages/web3/utils/compat/compat_stdlib.py", line 74, in sleep
    self.check()
  File "/home/user/.virtualenvs/prune/local/lib/python2.7/site-packages/web3/utils/compat/compat_stdlib.py", line 67, in check
    raise self
Timeout: 10 seconds`
* What type of node you were connecting to
Parity

### How can it be fixed?
Add timeout parameter to IPCProvider class to be able to retry sooner, when a timeout occurs. Right know 10 seconds timeout is hardcoded:  https://github.com/pipermerriam/web3.py/blob/master/web3/providers/ipc.py

### PR DESCRIPTION
**Pull request by jevogel**
### What was wrong?

Issue #309 requested that a timeout parameter be added to the IPCProvider class to allow more rapid retrying after timeouts.

### How was it fixed?

Fixed #309 by adding a timeout parameter to the IPCProvider class.

#### Cute Animal Picture

His name is Buddha!
[via reddit](https://www.reddit.com/r/aww/comments/7sdp8r/his_name_is_buddha/)

### CODE REVIEW

File: web3/providers/ipc.py [hunk 1]
```diff
@@ -136,12 +136,14 @@ def get_default_ipc_path(testnet=False):
 class IPCProvider(JSONBaseProvider):
     _socket = None
 
-    def __init__(self, ipc_path=None, testnet=False, *args, **kwargs):
+    def __init__(self, ipc_path=None, testnet=False, timeout=10,
+                 *args, **kwargs):
```
Comment (carver) [inline web3/providers/ipc.py]: *nitpick*

Our style guide says that if a function doesn't fit, then each argument goes on its own line: https://github.com/pipermerriam/ethereum-dev-tactical-manual/blob/master/style-guide.md#function-signatures

Like:

```py
    def __init__(
        self,
        ipc_path=None,
        ...
        **kwargs):
```

**Comment by carver** [comment]
Thanks! Could you also take a quick pass at the docs?

File: web3/providers/ipc.py (hunk 1 above)
Comment (jevogel) [reply web3/providers/ipc.py]: Thanks for pointing that out, I'll make the change.

**Comment by jevogel** [comment]
I added the new parameter to the IPCProvider section of docs/providers.rst.

**Comment by carver** [comment]
Yup, that looks good :+1: Thanks!

**Pull request merged**